utf8.c

```diff
@@ -2175,44 +2175,142 @@ If C<e E<lt> s> or if the scan would end up past C<e>, it raises a UTF8 warning
+     * time spent isn't large either way */
 
-    while (s < e) {
-        Ptrdiff_t expected_byte_count = UTF8SKIP(s);
+#ifndef EBCDIC
+
+    if (e - s < 96)
 
-        if (UNLIKELY(e - s  < expected_byte_count)) {
-            goto warn_and_return;
+#endif
+
+    {   /* Count characters directly */
+        while (x < e) {
+            x += UTF8SKIP(x);
+            len++;
         }
 
-        len++;
-        s += expected_byte_count;
+        if (LIKELY(e == x)) {
+            return len;
+        }
+
+      warn_and_return:
+        if (ckWARN_d(WARN_UTF8)) {
+            if (PL_op)
+                Perl_warner(aTHX_ packWARN(WARN_UTF8),
+                             "%s in %s", unees, OP_DESC(PL_op));
+            else
+                Perl_warner(aTHX_ packWARN(WARN_UTF8), "%s", unees);
+        }
+
+        return x - s;
+    }
+
+#ifndef EBCDIC
+
+    /* Count continuations, word-at-a-time.
+     *
+     * We need to stop before the final start character in order to
+     * preserve the limited error checking that's always been done */
+    const U8 * e_limit = e - UTF8_MAXBYTES;
+
+    /* Points to the first byte >=x which is positioned at a word boundary.  If
+     * x is on a word boundary, it is x, otherwise it is to the next word. */
+    const U8 * partial_word_end = x + PERL_WORDSIZE * PERL_IS_SUBWORD_ADDR(x)
+                                    - (PTR2nat(x) & PERL_WORD_BOUNDARY_MASK);
+
+    /* Process up to a full word boundary. */
+    while (x < partial_word_end) {
+        const Size_t skip = UTF8SKIP(x);
+
+        continuations += skip - 1;
+        x += skip;
     }
 
-    if (LIKELY(e == s)) {
-        return len;
+    /* Adjust back down any overshoot */
+    continuations -= x - partial_word_end;
+    x = partial_word_end;
+
+    do { /* Process per-word */
+
+        /* The idea for counting continuation bytes came from
+         * http://www.daemonology.net/blog/2008-06-05-faster-utf8-strlen.html
+         * One thing it does that this doesn't is to prefetch the buffer
+         *      __builtin_prefetch(&s[256], 0, 0);
+         *
+         * A continuation byte has the upper 2 bits be '10', and the rest
+         * dont-cares.  The VARIANTS mask zeroes out all but the upper bit of
+         * each byte in the word.  That gets shifted to the byte's lowest bit,
+         * and 'anded' with the complement of the 2nd highest bit of the byte,
+         * which has also been shifted to that position.  Hence the bit in that
+         * position will be 1 iff the upper bit is 1 and the next one is 0.  We
+         * then use the same integer multiplcation and shifting that are used
+         * in variant_under_utf8_count() to count how many of those are set in
+         * the word. */
+
+        continuations += (((((* (PERL_UINTMAX_T *) x)
+                                            & PERL_VARIANTS_WORD_MASK) >> 7)
+                      & (((~ (* (PERL_UINTMAX_T *) x))) >> 6))
```

Comment:
the casts to `PERL_UINTMAX_T *` cast away const, which will warn with `-Wcast-qual`.

Reply:
cone